Sources/SmokeHTTPClient/HTTPInvocationClient.swift

```diff
@@ -0,0 +1,294 @@
+        invocationTraceContext: TraceContextType,
+        invocationDelegate: HandlerDelegateType,
+        invocationMetrics: HTTPClientInvocationMetrics? = nil) {
+            self.init(
+                httpClient: httpClient,
+                ownsHttpClients: false,
+                retryConfiguration: retryConfiguration,
+                retryOnErrorProvider: retryOnErrorProvider,
+                invocationAttributes: invocationAttributes,
+                invocationTraceContext: invocationTraceContext,
+                invocationDelegate: invocationDelegate,
+                invocationMetrics: invocationMetrics)
+    }
+    #endif
+
+    private init(
+        httpClient: HTTPOperationsClient,
+        ownsHttpClients: Bool,
+        retryConfiguration: HTTPClientRetryConfiguration = .default,
+        retryOnErrorProvider: @escaping (SmokeHTTPClient.HTTPClientError) -> Bool = { error in error.isRetriable() },
+        invocationAttributes: HTTPClientInvocationAttributes,
+        invocationTraceContext: TraceContextType,
+        invocationDelegate: HandlerDelegateType,
+        invocationMetrics: HTTPClientInvocationMetrics? = nil) {
+            self.httpClient = httpClient
+
+            self.ownsHttpClients = ownsHttpClients
+            self.retryConfiguration = retryConfiguration
+            self.retryOnErrorProvider = retryOnErrorProvider
+
+            let reporting = StandardHTTPClientInvocationReporting(
+                internalRequestId: invocationAttributes.internalRequestId,
+                traceContext: invocationTraceContext,
+                logger: invocationAttributes.logger,
+                eventLoop: invocationAttributes.eventLoop,
+                outwardsRequestAggregator: invocationAttributes.outwardsRequestAggregator,
+                successCounter: invocationMetrics?.successCounter,
+                failure5XXCounter: invocationMetrics?.failure5XXCounter,
+                failure4XXCounter: invocationMetrics?.failure4XXCounter,
+                retryCountRecorder: invocationMetrics?.retryCountRecorder,
+                latencyTimer: invocationMetrics?.latencyTimer)
+
+            self.invocationContext = HTTPClientInvocationContext(reporting: reporting, handlerDelegate: invocationDelegate)
+    }
+
+    /**
+     Gracefully shuts down this client. This function is idempotent and
+     will handle being called multiple times. Will return when shutdown is complete.
+     */
+    public func shutdown() async throws {
+        if self.ownsHttpClients {
+            try await self.httpClient.shutdown()
+        }
+    }
+
+    public func executeRetriableWithoutOutput<InputType: HTTPRequestInputProtocol>(
+        endpoint: URL? = nil,
+        endpointPath: String,
+        httpMethod: HTTPMethod,
+        operation: String? = nil,
+        input: InputType) async throws {
+            guard endpoint != nil || self.httpClient.endpointHostName != "" else {
```

Comment:
I would include a comment just to make it clear what conceptually we are detecting here.

Reply:
Added this here and in other similar places:
```
// Users can either specify a default endpoint for the `HTTPInvocationClient` instance
// or pass in an request-specific endpoint. Doing neither is an error.
```